### PR TITLE
Reorganize module versions and created v4

### PR DESCRIPTION
**Pull request by ethankan**
Given this module is public-facing, I want to organize this repo in a way that is more consistent with Go best practices.

- Moved the Version 2 module code into the `v2` directory.
- There is no point in keeping any code for the "original" version (v1) of the module. So it's basically just an empty module.
- Starting a Version 4 for the module.
    - Created a `v4` directory.
    - Moved the source code under a `mode` directory, which represents the actual package.
    - Moved the internal packages under the `internal` directory.
    - Removed the deprecated TLS client auth feature.
    - New helper functions for setting/deleting key-value pairs.
    - New helper function to publish data for Data Routing.
    - Moved "dummy mqttd" to a standalone package, so that it's not automatically included by users of the `mode` package.
- Version 3 is unchanged, so our gateway code will be unaffected. Obviously, we should migrate GW to v4 eventually.
- I also set up separate GitHub Actions workflows for v3 and v4.

A lot of files are involved in this PR, so maybe it's easier to just browse the branch https://github.com/moderepo/device-sdk-go/tree/v4-devel to get a sense of what the repo would look like.

### CODE REVIEW

File: .github/workflows/check-v4.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+name: check v4
```
Comment (shigmas) [inline .github/workflows/check-v4.yml:1]: If we can sync up the indentation w/ v3, it'd be easier to keep them in sync (w/o ignoring whitespace)

**Review by zakkie — APPROVED** [review]
thank you and sorry for the late check

File: .github/workflows/check-v4.yml (hunk 1 above)
Comment (ethankan) [reply .github/workflows/check-v4.yml:1]: Fixed.

**Review by shigmas — COMMENTED** [review]
lgtm

**Pull request merged**